---
title: 'Practical 03 SG: LD and Haplotype estimation'
author: "Carlos Moyano & Kleber Reyes"
date: "`r Sys.Date()`"
output:
  pdf_document:
    latex_engine: xelatex
editor_options:
  chunk_output_type: console
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
rm(list = ls(all.names = TRUE))

```


```{r load packages, echo=FALSE, message=FALSE, warning=FALSE}
library(rstudioapi)
library(genetics)
library(HardyWeinberg)
library(LDheatmap)
library(haplo.stats)
set.seed(2209)
current_path <- getActiveDocumentContext()$path 
setwd(dirname(current_path)) 
```

# Linkage Disequilibrium

files: https://www.cog-genomics.org/plink/1.9/formats#bim

## 2. Load the FOXP2.dat file into the R environment. How many individuals and how many SNPs are there in the database? What percentage of the data is missing?

```{r}
# load data
genotypeData <- read.table("FOXP2/FOXP2.dat", header = TRUE)
genotypeData <- genotypeData[, 2: ncol(genotypeData)]
```


```{r}
individuals <- nrow(genotypeData); individuals
SNPs <- ncol(genotypeData); SNPs 

perc.mis <- 100*sum(is.na(genotypeData))/(SNPs*individuals); perc.mis
```

We have 104 individuals and 543 SNPs (variants) in the dataset. No missing data.

## 3. Determine the genotype counts for each SNP, and depict all SNPs simultaeneously in a ternary plot, and comment on your result. For how many variants do you reject Hardy-Weinberg equilibrium using an ordinary chi-square test without continuity correction? (hint: you can read the .bim in R in order to determine the alleles of each SNP, and use function MakeCounts from the HardyWeinberg package to create a matrix of genotype counts). ‘

```{r}
# load data
variantsColnames <- c("chromosome", "SNP id", "position", "basepair coordinate", "Allele 1", "Allele 2")
variantsData <- read.table("FOXP2/FOXP2.bim", header = FALSE, col.names = variantsColnames)
```

```{r}
alleles.SNP <- paste(variantsData[,"Allele.1"], variantsData[,"Allele.2"], sep="/")

genotype.counts <- MakeCounts(genotypeData, alleles.SNP, sep="/")[,1:3]

HWTernaryPlot(genotype.counts)
```

It seems that the majority of the SNPs follows the HWE

```{r}

chisq.pvalues <- HWChisqStats(genotype.counts, pvalues = TRUE)

sum(chisq.pvalues < 0.05)

```

There are 33 SNPs whose p-values is below the 0.05 threshold. We cannot say statistically that these variants follows the HWE.

## 4. Using the function LD from the genetics package, compute the LD statistic D for the SNPs rs34684677 and rs2894715 of the database. Is there significant association between the alleles of these two SNPs? 

```{r}

rs34684677 <- genotypeData[, "rs34684677"]
rs2894715 <- genotypeData[, "rs2894715"]

rs34684677.g <- genotype(rs34684677,sep="/")
rs2894715.g <- genotype(rs2894715,sep="/")

LD(rs34684677.g,rs2894715.g)

```

We can notice that the estimated D is negative (-0.055), then we are on the repulsion case, this entails that probability of heterozygotes is higher than the expected under independence, the standarize D or D' has value near to 1 (0.999). As D' is close to 1 we can say that this variables have a high LD. Since p-value is below 0.05 we can reject the H0 that states that D=0 (D'=0 and cor=0) and confirm that estimated values for D, D' and correlation are valids.

## 5. Also compute the LD statistic D for the SNPs rs34684677 and rs998302 of the database. Is there significant association between these two SNPs? Is there any reason why rs998302 could have stronger or weaker correlation than rs2894715?

```{r}

rs34684677 <- genotypeData[, "rs34684677"]
rs998302 <- genotypeData[, "rs998302"]

rs34684677.g <- genotype(rs34684677,sep="/")
rs998302.g <- genotype(rs998302,sep="/")

LD(rs34684677.g,rs998302.g)

```

In this case p-value > 0.05, therefore, we cannot reject the null hypothesis and we have to say that D, D' and correlation are 0.

## 6. Given your previous estimate of D for SNPs rs34684677 and rs2894715, infer the haplotype frequencies. Which haplotype is the most common?

```{r}


```

## 7. Compute the LD statistics R2 for all the marker pairs in this data base, using the LD function of the packages genetics. Be prepared that this make take a few minutes. Also compute an alternative estimate of R2 obtained by using the PLINK program.

For this purpose you should:
* Download and install PLINK 1.90 from https://www.cog-genomics.org/plink2/
* Take care to store the files FOXP2.bim, FOXP2.fam and FOXP2.bed in a directory where PLINK can find them.
* Compute LD estimates with PLINK using plink --bfile FOXP2 --r2 --matrix --out FOXP2 This creates a file with extension FOXP2.ld that contains a matrix with all R2 statistics. Read this file into the R environment. Make a scatter plot for R’s LD estimates against PLINK’s LD estimates. Are they identical or do they at least correlate? What’s the difference between these two estimators? Which estimator would your prefer and why?


```{r}


```

## 8. Compute a distance matrix with the distance in base pairs between all possible pairs of SNPs, using the basepair position of each SNP given in the .bim file. Make a plot of R’s R2 statistics against the distance (expressed as the number of basepairs) between the markers. Comment on your results.

```{r}


```

## 9. Make an LD heatmap of the markers in this database, using the R2 statistic with the LD function. Make another heatmap obtained by filtering out all variants with a MAF below 0.35, and redoing the computations to obtain the R2 statistics in R. Can you explain any differences observed between the two heatmaps?

```{r}


```

## 10. Can you distinguish blocks of correlated markers in the area of the FOXP2 gene? How many blocks do you think that at least seem to exist?

```{r}


```

## 11. Simulate independent SNPs under the assumption of Hardy-Weinberg equilibrium, using R’s sample instruction sample(c("AA","AB","BB"),n,replace=TRUE,prob=c(p\*p,2\*p\*q,q\*q))). Simulate as many SNPs as you have in your database, and take care to match each SNP in your database with a simulated SNP that has the same sample size and allele frequency. Make an LD heatmap of the simulated SNPs, using R2 as your statistic. Compare the results with the LD heatmap of the FOXP2 region. What do you observe? State your conclusions.

```{r}


```

# Haplotype estimation

## Load data

```{r}
haploData <- read.table("APOE/APOE.dat", header = TRUE)
haploData <- haploData[, 2: ncol(haploData)]
```

## 2. How many individuals and how many SNPs are there in the database? What percentage of the data is missing?

```{r}
individuals <- nrow(haploData); individuals
SNPs <- ncol(haploData); SNPs 

perc.mis <- 100*sum(is.na(haploData))/(SNPs*individuals); perc.mis

```

## 3. Assuming all SNPs are bi-allelic, how many haplotypes can theoretically be found for this data set?

```{r}


```

## 4. Estimate haplotype frequencies using the haplo.stats package (set the minimum posterior probability to 0.001). How many haplotypes do you find? List the estimated probabilities in decreasing order. Which haplotype number is the most common?

```{r}


```

## 5. Is the haplotypic constitution of any of the individuals in the database ambiguous or uncertain? For how many? What is the most likely haplotypic constitution of individual NA20763? (identify the constitution by the corresponding haplotype numbers).

```{r}


```

## 6. Suppose we would delete polymorphism rs374311741 from the database prior to haplotype estimation. Would this affect the results obtained? Justify your answer.

```{r}


```

## 7. Remove all genetic variants that have a minor allele frequency below 0.10 from the database, and re-run haplo.em. How does this affect the number of haplotypes?

```{r}


```

## 8. We could consider the newly created haplotypes in our last run of haplo.em as the alleles of a new superlocus. Which is, under the assumption of Hardy-Weinberg equilibrium, the most likely genotype at this new locus? What is the probability of this genotype? Which genotype is the second most likely, and what is its probability?

```{r}


```